### PR TITLE
Unit Test(s) Migration

### PR DESCRIPTION
**Pull request by andhikayuana**
# Overview

Introducing new test framework for `litewallet-android` project. As for now we will use `JUnit4` and `MockK`. more details, will share the docs on slack later and also probably we can have sharing session for this.

## Background Problem

The existing codebase is relatively hard to maintain, since most of the code uses singleton everywhere and not using dependency inversion principle. The current tests use androidTest, which is actually for UI Test, it’s (not suitable for Unit Test (AFAIK) and if we run the test it will require more resources (need emulator). More consideration is, we also need to make the codebase not so old by fully migrating to kotlin in the future. That’s why we need to create a new test strategy, especially for Unit Test.

## Unit Test Migration

- [x] `com/litewallet/tools/util/BRConstantsTest.kt`
- [x] `com/litewallet/tools/sqlite/TransactionDataSourceTest.java`
- [x] `com/platform/kvstore/ReplicatedKVStoreTest.java`
- [x] `com/litewallet/tools/util/Bip39ReaderTest.java`

## CI/CD

Currently the CI/CD can't run the unit tests because of mallformed env file that placed on circleci, will revisit later

### CODE REVIEW

File: app/src/androidTest/java/com/litewallet/analytics/ConstantsTests.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
 
 import java.net.URI;
 
+@Deprecated
```
Comment (kcw-grunt) [inline app/src/androidTest/java/com/litewallet/analytics/ConstantsTests.java:18]: @andhikayuana  Trying to run this on a new AS on a new test Mac and it cannot access the test module. 
How do we make sure it is reproducable of all Litewallet devs?

Comment (andhikayuana) [reply app/src/androidTest/java/com/litewallet/analytics/ConstantsTests.java:18]: noted, I'll create a document for this

**Comment by andhikayuana** [comment]
will revisit this later, need to provide google-services.json

File: .circleci/config.yml [hunk 2]
```diff
@@ -2,25 +2,51 @@
...
+          command: "echo Hello, World! && ls"
+      - run:
+          name: "Default for gradle.properties"
+          command: "echo \"RELEASE_STORE_FILE=/\nRELEASE_STORE_PASSWORD=\nRELEASE_KEY_ALIAS=\nRELEASE_KEY_PASSWORD=\nandroid.useAndroidX=true\nandroid.enableJetifier=true\" >> gradle.properties && ls && cat gradle.properties"
+      - run:
+          name: "Initialize submodule"
+          command: "git submodule init && git submodule update --init --recursive"
+      - run:
+          name: "Copy google-services.json"
+          command: echo "$GOOGLE_SERVICES_JSON" | base64 --decode > app/google-services.json
```
Comment (andhikayuana) [inline .circleci/config.yml:29]: the error from circleci mallformed here, need recheck the base64 encoded from google-services.json

File: app/src/test/java/com/litewallet/example/ExampleTest.kt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import io.mockk.mockk
+import io.mockk.spyk
+import io.mockk.verify
+import io.mockk.verifyOrder
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class ExampleTest {
```
Comment (andhikayuana) [inline app/src/test/java/com/litewallet/example/ExampleTest.kt:14]: this is some example basic unit tests

cc @kcw-grunt @josikie

File: app/src/test/java/com/litewallet/tools/util/BRConstantsTest.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.litewallet.tools.util
+
+import com.breadwallet.tools.util.BRConstants
+import org.junit.Assert
+import org.junit.Assert.assertSame
+import org.junit.Test
+
+//TODO: migrate from [com.litewallet.analytics.ConstantsTests]
+class BRConstantsTest {
```
Comment (andhikayuana) [inline app/src/test/java/com/litewallet/tools/util/BRConstantsTest.kt:9]: need to fill, please let me know if have questions

File: app/src/test/java/com/litewallet/tools/util/Bip39ReaderTest.kt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+package com.litewallet.tools.util
+
+//TODO: pls fill this
+class Bip39ReaderTest
```
Comment (andhikayuana) [inline app/src/test/java/com/litewallet/tools/util/Bip39ReaderTest.kt:4]: need to fill, please let me know if have questions

File: app/src/test/java/com/litewallet/tools/sqlite/TransactionDataSourceTest.kt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package com.litewallet.tools.sqlite
+
+//TODO: pls fill this
+class TransactionDataSourceTest
```
Comment (andhikayuana) [inline app/src/test/java/com/litewallet/tools/sqlite/TransactionDataSourceTest.kt:4]: need to fill, please let me know if have questions

File: app/src/test/java/com/litewallet/tools/security/BRKeyStoreTest.kt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package com.litewallet.tools.security
+
+//TODO: pls fill this
+class BRKeyStoreTest
```
Comment (andhikayuana) [inline app/src/test/java/com/litewallet/tools/security/BRKeyStoreTest.kt:4]: need to fill, please let me know if have questions

File: app/src/test/java/com/platform/kvstore/ReplicatedKVStoreTest.kt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package com.platform.kvstore
+
+//TODO: pls fill this
+class ReplicatedKVStoreTest
```
Comment (andhikayuana) [inline app/src/test/java/com/platform/kvstore/ReplicatedKVStoreTest.kt:4]: need to fill, please let me know if have questions

**Comment by kcw-grunt** [comment]
Pedas sama Ny. Suharti @andhikayuana !

File: .circleci/config.yml (hunk 2 above)
Comment (kcw-grunt) [reply .circleci/config.yml:29]: @andhikayuana  yes we need to add a debug version to circleci TODO

File: app/src/test/java/com/litewallet/tools/util/BRConstantsTest.kt (hunk 4 above)
Comment (kcw-grunt) [reply app/src/test/java/com/litewallet/tools/util/BRConstantsTest.kt:9]: @andhikayuana may check in the values onto this PR

File: app/src/test/java/com/litewallet/currency/CurrencyTests.kt [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import io.mockk.verify
+import kotlinx.coroutines.delay
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Test
+
+class CurrencyTests {
+
+    @Test
+    fun `invoke CurrencyDataSource instance and getAllCurrencies, should return the correct number of currencies`() {
```
Comment (kcw-grunt) [inline app/src/test/java/com/litewallet/currency/CurrencyTests.kt:17]: @andhikayuana  these are my attempts at the tests and I need your comments to fix them.  They do not run correctly

File: app/src/test/java/com/litewallet/currency/CurrencyTests.kt [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        val context: Context = mockk(relaxed = true)
+        val currencyDataSource = CurrencyDataSource.getInstance(context)
+        assertEquals(currencyDataSource.allCurrencies.count(), 20)
+    }
+
+    @Test
+    fun `invoke CurrencyDataSource instance and open database, then should validate database was not null`() {
```
Comment (kcw-grunt) [inline app/src/test/java/com/litewallet/currency/CurrencyTests.kt:24]: @andhikayuana  these are my attempts at the tests and I need your comments to fix them.  They do not run correctly

File: app/src/test/java/com/litewallet/currency/CurrencyTests.kt [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        val context: Context = mockk(relaxed = true)
+        val currencyDataSource = CurrencyDataSource.getInstance(context)
+        val database: SQLiteDatabase = currencyDataSource.openDatabase()
+        assertNotNull(database)
+    }
+
+    @Test
+    fun `invoke CurrencyDataSource instance database, should return currency by ISO`() {
```
Comment (kcw-grunt) [inline app/src/test/java/com/litewallet/currency/CurrencyTests.kt:32]: @andhikayuana  these are my attempts at the tests and I need your comments to fix them.  They do not run correctly

File: app/src/test/java/com/litewallet/currency/CurrencyTests.kt [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            fiatCurrency.code
+            fiatCurrency.valuePerLitecoin
+        }
+    }
+}
+
+class FiatCurrency(val code: String, val valuePerLitecoin: Double)
+class Amount(val value: Double)
+
+class CurrencySystemUnderTest(
```
Comment (kcw-grunt) [inline app/src/test/java/com/litewallet/currency/CurrencyTests.kt:92]: @andhikayuana  does this make sense?

File: app/src/test/java/com/litewallet/tools/util/SeedWordTests.kt [hunk 13]
```diff
@@ -0,0 +1,89 @@
+package com.litewallet.tools.util
+
+
+class SeedWordTests {
+}
+
+//
+//
+//package com.litewallet;
```
Comment (kcw-grunt) [inline app/src/test/java/com/litewallet/tools/util/SeedWordTests.kt:9]: cc: @josikie , @andhikayuana 
This is just a place holder as techdebt

**Review by kcw-grunt — COMMENTED** [review]
Need a look from @andhikayuana

File: app/src/test/java/com/litewallet/currency/CurrencyTests.kt (hunk 12 above)
Comment (andhikayuana) [reply app/src/test/java/com/litewallet/currency/CurrencyTests.kt]: I just removed this, pls check [this commit](https://github.com/litecoin-foundation/litewallet-android/pull/287/commits/3479a2479eca561ddc7f3a8b750e497a9600b454)

**Review by kcw-grunt — APPROVED** [review]
Looks Better to Me. We will continue working on this.

**Pull request merged**